### PR TITLE
Convert to typescript

### PR DESCRIPTION
**Pull request by paullinator**
---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1202960830332174

### CODE REVIEW

**Review by samholmes — COMMENTED** [review]
So far all of the changes look good. I'll need to dig deeper on this PR on Monday. I'm well into the big commit.

File: src/swap/godex.ts [hunk 1]
```diff
@@ -185,7 +215,7 @@ export function makeGodexPlugin(opts: EdgeCorePluginOptions): EdgeSwapPlugin {
           reply.min_amount,
           request.toCurrencyCode
         )
-        if (lt(toNativeAmount, nativeMin)) {
+        if (lt(toNativeAmount, nativeMin) === true) {
```
Comment (samholmes) [inline src/swap/godex.ts]: More `=== true`

File: src/swap/letsexchange.ts [hunk 2]
```diff
@@ -169,9 +177,11 @@ export function makeLetsExchangePlugin(
       // Calculate the amounts:
       let fromAmount, fromNativeAmount, toNativeAmount, reply
       if (request.quoteFor === 'from') {
-        reply = await call(uri + 'info', request, {
-          params: quoteParams
-        })
+        reply = asInfoReply(
```
Comment (samholmes) [inline src/swap/letsexchange.ts:180]: Maybe annotate `reply: ReturnType<typeof asInfoReply>` at declaration?

File: src/swap/sideshift.ts [hunk 3]
```diff
@@ -312,10 +320,7 @@ export function makeSideshiftPlugin(
 ): EdgeSwapPlugin {
   const { io, initOptions } = opts
 
-  const api = createSideshiftApi(
-    SIDESHIFT_BASE_URL,
-    io.fetchCors != null || io.fetch
-  )
+  const api = createSideshiftApi(SIDESHIFT_BASE_URL, io.fetch)
```
Comment (samholmes) [inline src/swap/sideshift.ts:323]: Another `fetchCors` logic change

File: test/demo.ts [hunk 4]
```diff
@@ -37,7 +39,7 @@ async function showRate(
   }
 }
 
-showRate(edgeCorePlugins.coinbase, 'iso:USD', 'BTC')
+showRate(edgeCorePlugins.coinbase, 'iso:USD', 'BTC').catch(e => console.log(e))
```
Comment (samholmes) [inline test/demo.ts:42]: Perhaps use `console.error`?

**Review by samholmes — CHANGES_REQUESTED** [review]
I see that the odd  `=== true` conditions were undid on the last commit, but why were they in place to begin?

Minor changes aside from a definite change regarding an unintended `!=`.

File: src/rate/coingecko.ts [hunk 5]
```diff
@@ -1,9 +1,14 @@
...
+  EdgeCorePluginOptions,
+  EdgeRateHint,
+  EdgeRatePair,
+  EdgeRatePlugin
+} from 'edge-core-js/types'
 
 const asGeckoBulkUsdReply = asMap(asObject({ usd: asNumber }))
 
-const coinGeckoMap = {
+const coinGeckoMap: { [ccode: string]: string } = {
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/rate/coingecko.ts:11]: `ccode` is short for currencyCode

File: src/swap/sideshift.ts [hunk 6]
```diff
@@ -314,7 +312,10 @@ export function makeSideshiftPlugin(
 ): EdgeSwapPlugin {
   const { io, initOptions } = opts
 
-  const api = createSideshiftApi(SIDESHIFT_BASE_URL, io.fetchCors || io.fetch)
+  const api = createSideshiftApi(
+    SIDESHIFT_BASE_URL,
+    io.fetchCors != null || io.fetch
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/swap/sideshift.ts]: These should actually use fetchCors, not io.fetch. I'll fix it.

File: src/rate/coincap.ts [hunk 7]
```diff
@@ -31,7 +26,7 @@ const currencyMap = {}
 
 export function makeCoincapPlugin(opts: EdgeCorePluginOptions): EdgeRatePlugin {
   const { io, log } = opts
-  const fetch = io.fetchCors || io.fetch
+  const fetch = io.fetchCors != null || io.fetch
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/rate/coincap.ts]: These should actually use fetchCors, not io.fetch. I'll fix it.

File: src/swap/foxExchange.ts [hunk 8]
```diff
@@ -319,27 +338,27 @@ export function makeFoxExchangePlugin(
                 ? dummyTx.parentNetworkFee
                 : dummyTx.networkFee
           },
-          destinationAddress,
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/swap/foxExchange.ts:324]: only in commented out code in the action queue. This was moved to EdgeSwapResult vs EdgeSwapQuote

File: src/swap/changenow.ts [hunk 9]
```diff
@@ -193,15 +193,18 @@ export function makeChangeNowPlugin(
 
         const { minAmount, maxAmount } = asMarketRange(marketRangeResponseJson)
 
-        if (lt(largeDenomAmount, minAmount.toString())) {
+        if (lt(largeDenomAmount, minAmount.toString()) === true) {
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/swap/changenow.ts]: Automation required it in one of the commits, then after some changes it auto removed the `=== true`. I believe it had to do with VSCode needing a reboot. I've found VSCode won't have up to date squiggly errors after a change to packages and rules until a reboot

File: src/swap/changenow.ts [hunk 10]
```diff
@@ -193,15 +193,18 @@ export function makeChangeNowPlugin(
...
             minAmount.toString(),
             fromCurrencyCode
           )
           throw new SwapBelowLimitError(swapInfo, minNativeAmount)
         }
 
-        if (maxAmount != null && gt(largeDenomAmount, maxAmount.toString())) {
+        if (
+          maxAmount != null &&
+          gt(largeDenomAmount, maxAmount.toString()) === true
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/swap/changenow.ts]: same as above

File: src/swap/defi/thorchain.ts [hunk 11]
```diff
@@ -72,14 +72,14 @@ const EVM_CURRENCY_CODES = {
...
-const MAINNET_CODE_TRANSCRIPTION = {
+const MAINNET_CODE_TRANSCRIPTION: { [cc: string]: ChainTypes } = {
   bitcoin: 'BTC',
   bitcoincash: 'BCH',
   binancechain: 'BNB',
   litecoin: 'LTC',
   ethereum: 'ETH',
   dogecoin: 'DOGE',
-  thorchain: 'RUNE'
+  thorchain: 'THOR'
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/swap/defi/thorchain.ts:82]: That's the correct name of the chain. We dont yet support the chain so it was not a bug 'yet'

File: src/swap/defi/thorchain.ts [hunk 12]
```diff
@@ -191,14 +190,14 @@ export function makeThorchainPlugin(
             fetchInfo(fetch, 'v1/exchangeInfo/edge')
           )
 
-          if (exchangeInfoResponse.ok) {
+          if (exchangeInfoResponse.ok === true) {
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/swap/defi/thorchain.ts:193]: same as comment above

File: src/swap/defi/thorchain.ts [hunk 13]
```diff
@@ -240,13 +239,13 @@ export function makeThorchainPlugin(
         })
       ])
 
-      if (!iaResponse.ok) {
+      if (iaResponse.ok === false) {
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/swap/defi/thorchain.ts:242]: yep

File: src/swap/defi/uni-v2-based/uniV2Utils.ts [hunk 14]
```diff
@@ -34,7 +33,7 @@ export const getSwapAmounts = async (
     : []
   ).map(String)
 
-  if (!amountToSwap || !expectedAmountOut)
+  if (amountToSwap == null || expectedAmountOut != null)
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/swap/defi/uni-v2-based/uniV2Utils.ts:36]: Good catch

File: src/swap/defi/uni-v2-based/uniV2Utils.ts [hunk 15]
```diff
@@ -102,12 +101,12 @@ export const getSwapTransactions = async (
...
 
   const txs = await (async (): Promise<
-    Array<Promise<PopulatedTransaction> | undefined>
+    Array<Promise<PopulatedTransaction | undefined>>
   > => {
     // Deposit native currency for wrapped token
     if (isFromNativeCurrency && isToWrappedCurrency) {
       return [
-        makeWrappedFtmContract.populateTransaction.deposit({
+        makeWrappedFtmContract(provider).populateTransaction.deposit({
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/swap/defi/uni-v2-based/uniV2Utils.ts:109]: As Jon. I have NO idea how. I brought this up w/him. But I verified it works now

File: src/swap/defi/uni-v2-based/uniV2Utils.ts [hunk 16]
```diff
@@ -175,7 +177,16 @@ export const getSwapTransactions = async (
...
+  const out: ethers.PopulatedTransaction[] = []
+
+  for (const promiseTx of txs) {
+    const tx = await promiseTx
+    if (tx != null) {
+      const cleanTx: ethers.PopulatedTransaction = tx
+      out.push(cleanTx)
+    }
+  }
+  return out
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/swap/defi/uni-v2-based/uniV2Utils.ts:189]: They were already serial with the awaits that were in the array.

File: src/rate/coincap.ts [hunk 17]
```diff
@@ -22,20 +22,26 @@ const asCoincapAssets = asArray(
...
+  toCurrency: string
+  rate: number
+}
+
+const currencyMap: { [symbol: string]: string } = {}
 
 export function makeCoincapPlugin(opts: EdgeCorePluginOptions): EdgeRatePlugin {
   const { io, log } = opts
-  const fetch = io.fetchCors != null || io.fetch
+  const { fetch } = io
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/rate/coincap.ts:35]: These should actually use fetchCors, not io.fetch. I'll fix it.

**Review by samholmes — APPROVED** [review]

**Pull request merged**